File: sdks/typescript/src/notifications.ts
/**
 * Notifications Client
 */

import { HttpClient } from './http';
import {
  Notification,
  NotificationPreferences,
  PaginationParams,
  PaginatedResponse,
  NotificationType,
} from './types';

export class NotificationsClient {
  constructor(private http: HttpClient) {}

  /**
   * List notifications
   */
  async list(
    params?: PaginationParams & { unreadOnly?: boolean; type?: NotificationType }
  ): Promise<PaginatedResponse<Notification>> {
    const response = await this.http.get<PaginatedResponse<any>>(
      '/notifications',
      { params }
    );

    return {
      ...response,
      data: response.data.map(this.transformNotification),
    };
  }

  /**
   * Get notification by ID
   */
  async get(notificationId: string): Promise<Notification> {
    const response = await this.http.get<any>(`/notifications/${notificationId}`);
    return this.transformNotification(response);
  }

  /**
   * Mark notification as read
   */
  async markAsRead(notificationId: string): Promise<void> {
    await this.http.post(`/notifications/${notificationId}/read`);
  }

  /**
   * Mark all notifications as read
   */
  async markAllAsRead(): Promise<void> {
    await this.http.post('/notifications/read-all');
  }

  /**
   * Delete notification
   */
  async delete(notificationId: string): Promise<void> {
    await this.http.delete(`/notifications/${notificationId}`);
  }

  /**
   * Delete all notifications
   */
  async deleteAll(): Promise<void> {
    await this.http.delete('/notifications');
  }

  /**
   * Get unread count
   */
  async getUnreadCount(): Promise<number> {
    const response = await this.http.get<{ count: number }>('/notifications/unread-count');
    return response.count;
  }

  /**
   * Get notification preferences
   */
  async getPreferences(): Promise<NotificationPreferences> {
    return this.http.get<NotificationPreferences>('/notifications/preferences');
  }

  /**
   * Update notification preferences
   */
  async updatePreferences(
    preferences: Partial<NotificationPreferences>
  ): Promise<NotificationPreferences> {
    return this.http.put<NotificationPreferences>(
      '/notifications/preferences',
      preferences
    );
  }

  /**
   * Subscribe to push notifications
   */
  async subscribePush(subscription: PushSubscriptionJSON): Promise<void> {
    await this.http.post('/notifications/push/subscribe', subscription);
  }

  /**
   * Unsubscribe from push notifications
   */
  async unsubscribePush(endpoint: string): Promise<void> {
    await this.http.post('/notifications/push/unsubscribe', { endpoint });
  }

  /**
   * Test notification delivery
   */
  async sendTest(channel: 'email' | 'push' | 'inApp'): Promise<void> {
    await this.http.post('/notifications/test', { channel });
  }

  /**
   * Get notification history
   */
  async getHistory(
    params?: PaginationParams & {
      startDate?: Date;
      endDate?: Date;
      type?: NotificationType;
    }
  ): Promise<PaginatedResponse<Notification>> {
    const response = await this.http.get<PaginatedResponse<any>>(
      '/notifications/history',
      {
        params: {
          ...params,
          startDate: params?.startDate?.toISOString(),
          endDate: params?.endDate?.toISOString(),
        },
      }
    );

    return {
      ...response,
      data: response.data.map(this.transformNotification),
    };
  }

  /**
   * Transform raw notification response
   */
  private transformNotification(raw: any): Notification {
    return {
      ...raw,
      createdAt: new Date(raw.createdAt),
    };
  }
}

// WebSocket connection for real-time notifications
export class NotificationsWebSocket {
  private ws: WebSocket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private handlers: Map<string, Set<(notification: Notification) => void>> = new Map();

  constructor(
    private url: string,
    private token: string
  ) {}

  /**
   * Connect to notifications WebSocket
   */
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(`${this.url}?token=${this.token}`);

        this.ws.onopen = () => {
          this.reconnectAttempts = 0;
          resolve();
        };

        this.ws.onmessage = (event) => {
          try {
            const data = JSON.parse(event.data);
            this.handleMessage(data);
          } catch (error) {
            console.error('Failed to parse notification:', error);
          }
        };

        this.ws.onerror = (error) => {
          reject(error);
        };

        this.ws.onclose = () => {
          this.handleDisconnect();
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Subscribe to notification type
   */
  on(type: NotificationType | '*', handler: (notification: Notification) => void): () => void {
    const key = type;
    if (!this.handlers.has(key)) {
      this.handlers.set(key, new Set());
    }
    this.handlers.get(key)!.add(handler);

    // Return unsubscribe function
    return () => {
      this.handlers.get(key)?.delete(handler);
    };
  }

  /**
   * Handle incoming message
   */
  private handleMessage(data: any): void {
    const notification: Notification = {
      ...data,
      createdAt: new Date(data.createdAt),
    };

    // Call type-specific handlers
    const typeHandlers = this.handlers.get(notification.type);
    if (typeHandlers) {
      typeHandlers.forEach((handler) => handler(notification));
    }

    // Call wildcard handlers
    const wildcardHandlers = this.handlers.get('*');
    if (wildcardHandlers) {
      wildcardHandlers.forEach((handler) => handler(notification));
    }
  }

  /**
   * Handle disconnection with reconnect
   */
  private handleDisconnect(): void {
    if (this.reconnectAttempts < this.maxReconnectAttempts) {
      this.reconnectAttempts++;
      const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

      setTimeout(() => {
        this.connect().catch((error) => {
          console.error('Reconnection failed:', error);
        });
      }, delay);
    }
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}

File: sdks/python/fileops/notifications.py
"""
Notifications Client
"""

from typing import Optional
from datetime import datetime
import httpx
from .types import Notification, NotificationPreferences, PaginatedResponse


class NotificationsClient:
    """Notifications management client."""

    def __init__(self, http: httpx.Client):
        self._http = http

    def list(
        self,
        unread_only: bool = False,
        type: Optional[str] = None,
        page: int = 1,
        limit: int = 20,
    ) -> PaginatedResponse:
        """List notifications."""
        params = {"page": page, "limit": limit, "unreadOnly": unread_only}
        if type:
            params["type"] = type

        response = self._http.get("/notifications", params=params)
        response.raise_for_status()
        data = response.json()
        return PaginatedResponse(
            data=[self._parse_notification(n) for n in data["data"]],
            page=data["pagination"]["page"],
            limit=data["pagination"]["limit"],
            total=data["pagination"]["total"],
            total_pages=data["pagination"]["totalPages"],
            has_next=data["pagination"]["hasNext"],
            has_prev=data["pagination"]["hasPrev"],
        )

    def get(self, notification_id: str) -> Notification:
        """Get notification by ID."""
        response = self._http.get(f"/notifications/{notification_id}")
        response.raise_for_status()
        return self._parse_notification(response.json())

    def mark_as_read(self, notification_id: str) -> None:
        """Mark notification as read."""
        response = self._http.post(f"/notifications/{notification_id}/read")
        response.raise_for_status()

    def mark_all_as_read(self) -> None:
        """Mark all notifications as read."""
        response = self._http.post("/notifications/read-all")
        response.raise_for_status()

    def delete(self, notification_id: str) -> None:
        """Delete notification."""
        response = self._http.delete(f"/notifications/{notification_id}")
        response.raise_for_status()

    def delete_all(self) -> None:
        """Delete all notifications."""
        response = self._http.delete("/notifications")
        response.raise_for_status()

    def get_unread_count(self) -> int:
        """Get unread notification count."""
        response = self._http.get("/notifications/unread-count")
        response.raise_for_status()
        return response.json()["count"]

    def get_preferences(self) -> NotificationPreferences:
        """Get notification preferences."""
        response = self._http.get("/notifications/preferences")
        response.raise_for_status()
        data = response.json()
        return NotificationPreferences(
            email=data["email"],
            push=data["push"],
            in_app=data["inApp"],
            types=data.get("types", {}),
        )

    def update_preferences(
        self,
        email: Optional[bool] = None,
        push: Optional[bool] = None,
        in_app: Optional[bool] = None,
        types: Optional[dict] = None,
    ) -> NotificationPreferences:
        """Update notification preferences."""
        payload = {}
        if email is not None:
            payload["email"] = email
        if push is not None:
            payload["push"] = push
        if in_app is not None:
            payload["inApp"] = in_app
        if types is not None:
            payload["types"] = types

        response = self._http.put("/notifications/preferences", json=payload)
        response.raise_for_status()
        data = response.json()
        return NotificationPreferences(
            email=data["email"],
            push=data["push"],
            in_app=data["inApp"],
            types=data.get("types", {}),
        )

    def _parse_notification(self, data: dict) -> Notification:
        return Notification(
            id=data["id"],
            type=data["type"],
            title=data["title"],
            message=data["message"],
            read=data["read"],
            data=data.get("data"),
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00"))
            if data.get("createdAt")
            else None,
        )
